scan-external-contacts/nodejs/index.js

```diff
@@ -0,0 +1,296 @@
+// Import built in libraries needed.
+const fs = require('fs');
+const csv = require('fast-csv');
+
+// Set Genesys Cloud objects
+const platformClient = require('purecloud-platform-client-v2');
+const apiClient = platformClient.ApiClient.instance;
+// API Instances that will be used to fetch external contacts, organizations, notes and relationships from Genesys Cloud
+const contactsApi = new platformClient.ExternalContactsApi();
+
+// OAuth Client Grant Credentials
+const apiClientId = '-- id here --';
+const apiClientSecret = '-- secret here --';
+const apiClientRegion = '-- genesys cloud region here - ex: mypurecloud.com --';
+
+// Script Sample Settings
+// - Maximum number of records we will request per "page" when using the External Contacts Scan API endpoints (10 to 200)
+const maxRecordsPerPage = 100;
+// - Force the script to make a pause between 2 consecutive API requests (to control/prevent API Rate limit)
+// i.e. Minimum time between 2 API requests
+const minAPIRequestInterval = 500;
+// i.e. If there is a pause to make (processing time < minAPIRequestInterval), the pause will be at minimum minWaitTime
+const minWaitTime = 50;
+// - Define a maximum number of retries (for API Rate Limit rejects)
+const maxRateLimitRetryAttempts = 2;
+// - Settings for Export to file (CSV or RAW)
+const fieldSeparator = ',';
+const exportType = 'CSV';
+
+// Global vars
+var writeStream, csvStream;
+var consecutiveRateLimitRetryAttempts = 0;
+
+/*
+------------------------------------------------------------------------------
+* Login, Scan Objects
+------------------------------------------------------------------------------
+*/
+
+//#region Login, Scan
+
+// Login using Client Credentials Grant (returns: true (success), false (failure))
+const login = async () => {
+
+    try {
+        apiClient.setEnvironment(apiClientRegion);
+
+        let creds = await apiClient.loginClientCredentialsGrant(apiClientId, apiClientSecret);
+
+        console.log('Login Success. Token will expire at: ', creds.tokenExpiryTime);
+
+        return true;
+    }
+    catch (error) {
+        console.error('Login error - Error:', JSON.stringify(error, null, 2));
+
+        return false;
+    }
+}
+
+// Perform an External Contacts Scan API request and manage errors (429)
+const scanContacts = async (cursor) => {
+
+    // Defining the default values for the result structure
+    let result = {
+        error: false,
+        errorMsg: "",
+        entities: [],
+        isEmpty: true,
+        isFinished: true,
+        nextCursor: null,
+        startTime: 0
+    };
+
+    try {
+        result.startTime = Date.now();
+
+        let opts = {
+            'limit': maxRecordsPerPage
+        };
+        // No opts.cursor on first scan - cursor == null
+        if (cursor) {
+            opts.cursor = cursor;
+        }
+
+        // External Contacts Scan API request
+        let scanData = await contactsApi.getExternalcontactsScanContacts(opts);
```

Comment:
I would recommend you take a look at https://github.com/MyPureCloud/user-provisioning-scripts-js/blob/master/src/proxies/groupsapi.js.  I have got a pretty good pattern for how to assemble pages of data into a single data structure.  The key thing is really understanding how much data you are bringing back.  If it massive, you will run of memory, but most calls are going to be small in size with a few pages of data being retrieved.  The code tends to be more readable and then you can process the data as a single flate data structure.  Your approach works but it is very verbose.